### PR TITLE
feat: add --external-links flag for cross-origin directory links

### PR DESCRIPTION
**Pull request by fangfufu**
This pull request introduces opt-in support for mounting and caching external (cross-origin) directory links, resolving #135. It ensures robust origin comparison, secure caching behavior, same-origin protections for HTTP credentials and custom headers, and comprehensive unit and integration testing.

#### 1. Core External Links Flag (`--external-links`)
- Allows the HTML parser to detect `<a href>` tags pointing to different origins, extracts the filename, and exposes those files and directories alongside the local directory listing.
- **First-Wins Deduplication**: If multiple external links produce the same filename, the first one encountered is retained.
- **Recursive Traversal**: External links ending with `/` are treated as directories; navigating into them triggers a recursive directory traversal on the remote URL.
- Skips URL reconstruction in `LinkTable_fill()` for pre-populated external URLs.

#### 2. Security Hardening
- **Origin Comparisons**: Restricts origin comparisons using strict host and case-insensitive scheme matching (per RFC 3986) to identify cross-origin links.
- **Credential & Header Isolation**: Custom headers (`-H`) and HTTP credentials (`-u`/`-p`) are strictly scoped to the mounted (root) server origin using a new `is_same_origin()` helper. They are not sent to cross-origin external links.
- **Buffer Safety**: Added boundary checks for pointer arithmetic when handling same-origin URLs shorter than `ROOT_LINK_OFFSET`. Replaced insecure `strncpy` with safe `snprintf` in `LinkTable_fill` to guarantee null-termination.
- **Null Safety**: Defensive `NULL` checks and graceful fallbacks around `curl_easy_escape`, `curl_easy_unescape`, and related string utilities to prevent pointer dereferences.
- **Path Traversal Mitigation**: Sanitizes unescaped paths in `LinkTable_new` to prevent path traversal via `..` for external links.

#### 3. Caching Support for External Links
- Refactored the cache system to support external links without relying on the hardcoded `ROOT_LINK_OFFSET`.
- Implemented `url_to_cache_path()` in `src/link.c` to safely generate unescaped, sanitized cache paths.
- Updated `Cache_create`, `Cache_open`, and `Cache_delete` to utilize the new cache path mapping.
- Resolved a duplicate clang-tidy warning in `Cache_create` by consolidating `SONIC` and `SINGLE` modes.

### Testing & Validation
- **Unit Tests**: Added 25 unit test cases in `tests/test_link.c` covering URL scheme parsing, host casing comparisons, first-wins deduplication, query/fragment extraction, and edge-case inputs.
- **Integration Tests**: Added 5 end-to-end integration tests in `tests/integration/run_integration_test.sh` utilizing a second ephemeral HTTP server as the external origin.

## Summary by CodeRabbit

* **New Features**
  * Added --external-links to include external cross-origin links in directory listings; duplicate filenames deduplicated (first-wins).

* **Documentation**
  * Help/usage updated to document --external-links (off by default).

* **Behavior**
  * External links are exposed and traversable when enabled; HTTP credentials are applied only to the mounted server and cross-origin targets that need auth emit a warning. Cache naming and sanitization for external URLs improved.

* **Tests**
  * Expanded unit and integration tests covering parsing, cross-origin detection, deduplication, traversal, and cache-path behavior.

[Review Change Stack](https://app.coderabbit.ai/change-stack/fangfufu/httpdirfs/pull/279?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/link.c [hunk 1]
```diff
@@ -816,10 +1059,46 @@ LinkTable *LinkTable_alloc(const char *url)
...
+        if (temp) {
+            curl_free(temp);
+        }
+        /* Sanitize unescaped_path to prevent path traversal via ".." */
+        char *p = unescaped_path;
+        while ((p = strstr(p, ".."))) {
+            p[0] = '_';
+            p[1] = '_';
+            p += 2;
+        }
```
[the remark this answers is not in the document]
Comment (fangfufu) [reply src/link.c:1096]: Fixed by sanitizing the unescaped path to replace `/` and `:` directory separators/colons with `_`. Added a unit test to verify external URL cache path flattening/sanitization.

File: src/link.c [hunk 2]
```diff
@@ -816,10 +1059,46 @@ LinkTable *LinkTable_alloc(const char *url)
     return linktbl;
 }
 
-LinkTable *LinkTable_new(const char *url)
+char *url_to_cache_path(const char *url)
 {
     char *unescaped_path;
```
[the remark this answers is not in the document]
Comment (fangfufu) [reply src/link.c:1076]: Fixed by adding a defensive NULL check at the beginning of `url_to_cache_path()`. Added a unit test to verify NULL parameter handling.
